tests/goth/test_run_yacat.py

```diff
@@ -0,0 +1,141 @@
+            remaining_tasks.discard(task_data)
+        if not remaining_tasks:
+            return
+
+    raise AssertionError(f"Tasks not {status}: {remaining_tasks}")
+
+
+async def assert_all_tasks_sent(output_lines: EventStream[str]):
+    """Assert that for every task a line with `Task sent` will appear."""
+    await assert_all_tasks_processed("sent", output_lines)
+
+
+async def assert_all_tasks_computed(output_lines: EventStream[str]):
+    """Assert that for every task a line with `Task computed` will appear."""
+    await assert_all_tasks_processed("computed", output_lines)
+
+
+async def assert_all_invoices_accepted(output_lines: EventStream[str]):
+    """Assert that an invoice is accepted for every provider that confirmed an agreement."""
+    unpaid_agreement_providers = set()
+
+    async for line in output_lines:
+        m = re.search("Agreement confirmed by provider '([^']*)'", line)
+        if m:
+            prov_name = m.group(1)
+            logger.debug("assert_all_invoices_accepted: adding provider '%s'", prov_name)
+            unpaid_agreement_providers.add(prov_name)
+        m = re.search("Accepted invoice from '([^']*)'", line)
+        if m:
+            prov_name = m.group(1)
+            logger.debug("assert_all_invoices_accepted: adding invoice for '%s'", prov_name)
+            unpaid_agreement_providers.remove(prov_name)
+
+    if unpaid_agreement_providers:
+        raise AssertionError(f"Unpaid agreements for: {','.join(unpaid_agreement_providers)}")
+
+
+@pytest.mark.asyncio
+async def test_run_yacat(
+    log_dir: Path,
+    project_dir: Path,
+) -> None:
+
+    # This is the default configuration with 2 wasm/VM providers
+    goth_config = load_yaml(Path(__file__).parent / "assets" / "goth-config.yml")
+
+    yacat_path = project_dir / "examples" / "yacat" / "yacat.py"
+
+    configure_logging(log_dir)
+
+    runner = Runner(
+        base_log_dir=log_dir,
+        compose_config=goth_config.compose_config,
+    )
+
+    async with runner(goth_config.containers):
+
+        requestor = runner.get_probes(probe_type=RequestorProbe)[0]
+
+        async with requestor.run_command_on_host(
+            f"{yacat_path} ?a?a $P$5ZDzPE45CigTC6EY4cXbyJSLj/pGee0 --number-of-providers 2 --log-file yacat-debug.log  --subnet-tag goth",
+            env=os.environ,
+        ) as (_cmd_task, cmd_monitor):
+
+            # Add assertions to the command output monitor `cmd_monitor`:
+            cmd_monitor.add_assertion(assert_no_errors)
+            cmd_monitor.add_assertion(assert_keyspace_found)
+            cmd_monitor.add_assertion(assert_password_found)
+            cmd_monitor.add_assertion(assert_all_invoices_accepted)
+            all_sent = cmd_monitor.add_assertion(assert_all_tasks_sent)
+            all_computed = cmd_monitor.add_assertion(assert_all_tasks_computed)
+
+            await cmd_monitor.wait_for_pattern(".*Received proposals from 2 ", timeout=20)
+            logger.info("Received proposals")
+
```

Comment:
I'd replace the assertions `assert_keyspace_found` and `assert_password_found` with appropriate `wait_for_pattern()` calls -- I think we know in what order they should appear, with respect to other patterns we're waiting for in this test.